File: src/Layout/Widgets/Features/ConfiguringFeature.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

package Layout.Widgets.Features;

import Configuration.Config;
import Configuration.Configurable;
import java.util.Collection;

/**
 * Editor for {@link Configurable}.
 * 
 * @author Plutonium_
 */
@Feature(
  name = "Configurator",
  description = "Provides settings and configurations",
  type = ConfiguringFeature.class
)
public interface ConfiguringFeature {
    
    /** Displays configs of the specified configurable object for user to edit. */
    default void configure(Configurable c) {
        configure(c.getFields());
    };
    
    /** Displays specified configs for user to edit. */
    void configure(Collection<Config> c);
}

File: src/sp/it/pl/layout/widget/feature/ConfiguringFeature.kt
package sp.it.pl.layout.widget.feature

import sp.it.pl.util.conf.Config
import sp.it.pl.util.conf.Configurable

/** Editor for [Configurable]. */
@Feature(name = "Configurator", description = "Provides settings and configurations", type = ConfiguringFeature::class)
interface ConfiguringFeature<T> {

    /** Display configs of the specified configurable object for user to edit. */
    fun configure(configurable: Configurable<out T>?) = configure(configurable?.fields ?: listOf())

    /** Display specified configs for user to edit. */
    fun configure(configurable: Collection<Config<out T>>)

}